File: dbc_talks/linked_lists/linked_list.rb
Node = Struct.new(:value, :successor) do
  def self.build(args)
    self.new(args[:value], args[:successor])
  end
end

List = Struct.new(:head) do
  def self.build(args)
    self.new(args[:head])
  end

  def each
    node = self.head
    while node
      yield node
      node = node.successor
    end
  end

  def find(value)
    find_node(node: self.head, value: value)
  end

  def find_node(args)
    node, value = args[:node], args[:value]
    if !node || node.value == value
      return node
    else
      args[:node] = node.successor
      find_node(args)
    end
  end

  def unshift(value)
    self.head = Node.build(value: value, successor: self.head)
  end

  def push(value)
    self.tail.successor = Node.build(value: value)
  end

  def tail
    self.each{|node| return node if node.successor.nil?}
  end

end

c = Node.build(value: 47)
b = Node.build(value: 23, successor: c)
a = Node.build(value: 'original head', successor: b)

list = List.build(head: a)

puts list.head == a
puts list.head.successor == b
puts list.find('original head') == a
puts list.find(47) == c
puts list.find(47) == list.tail
puts list.find(46) == nil

list.unshift('new head')
list.push('new tail')

puts list.find('new tail') == list.tail

puts list.find('new head') == list.head

File: dbc_talks/linked_lists/linked_list.js
var Node = function(value, successor) {
  this.value = value;
  this.successor = successor;
}

var List = function(head) {
  this.head = head;
}

List.prototype.each = function(callback) {
  var node = this.head;
  while (node) {
    callback(node);
    node = node.successor;
  }
}

List.prototype.push = function(value) {
  this.tail().successor = new Node(value);
}

List.prototype.unshift = function(value) {
  return this.head = new Node(value, this.head);
}

List.prototype.tail = function() {
  var tail
  this.each(function(node) {
    if (node.successor === undefined) {
      tail = node;
    }
  })
  return tail;
}

List.prototype.find = function(value) {
  return this.find_node(this.head, value)
}

List.prototype.find_node = function(node, value) {
  if ( !node || value === node.value) {
    return node;
  }
  else {
    return this.find_node(node.successor, value)
  }
}

List.prototype.find_all = function(value) {
  var matches = [];
  this.each(function(node) {
    if (node.value === value) {
      matches.push(node);
    }
  });
  return matches;
}

c = new Node(23);
b = new Node(47, c);
a = new Node('original head', b);

list = new List(a);

list.each(function(node) {
  console.log(node.value);
});

console.log(list.find(47) === b);
console.log(list.find_all(47).length === 1);
console.log(list.find_all(23)[0] === c);

list.push(47);
x = list.unshift('new head');

console.log(list.find_all(47).length === 2);
console.log(list.find_all(47)[0] === list.find(47));
console.log(list.head === x);